File: sfgame-web/src/core/batch.ts
/**
 * 即时模式三角形网格构建器（纯计算，无 DOM/WebGL 依赖，可无头测试）。
 * 顶点格式 x,y,r,g,b,a（颜色 0..1，非预乘 alpha）平铺于 Float32Array，渲染层只负责
 * 把 data.subarray(0, count * VERTEX_STRIDE) 以 TRIANGLES 上传绘制。逐顶点着色使每帧
 * 一次 draw call、图元可带精确透明度，线段宽度也不受 GL lineWidth 恒为 1 的限制。
 */
export const VERTEX_STRIDE = 6

/** 椭圆/圆弧展开的默认分段数（世界尺度下半径 0.3~12，20 段已足够圆滑） */
export const DISC_SEGMENTS = 20

export class MeshBatch {
  data: Float32Array
  count = 0

  constructor(capacity = 32768) {
    this.data = new Float32Array(capacity * VERTEX_STRIDE)
  }

  reset() {
    this.count = 0
  }

  private ensure(extra: number) {
    const need = (this.count + extra) * VERTEX_STRIDE
    if (need <= this.data.length) return
    let cap = this.data.length
    while (cap < need) cap *= 2
    const next = new Float32Array(cap)
    next.set(this.data)
    this.data = next
  }

  private push(x: number, y: number, r: number, g: number, b: number, a: number) {
    const o = this.count * VERTEX_STRIDE
    const d = this.data
    d[o] = x
    d[o + 1] = y
    d[o + 2] = r
    d[o + 3] = g
    d[o + 4] = b
    d[o + 5] = a
    this.count++
  }

  tri(
    x0: number, y0: number,
    x1: number, y1: number,
    x2: number, y2: number,
    r: number, g: number, b: number, a: number,
  ) {
    this.ensure(3)
    this.push(x0, y0, r, g, b, a)
    this.push(x1, y1, r, g, b, a)
    this.push(x2, y2, r, g, b, a)
  }

  /** 轴对齐矩形，y0 < y1 */
  rect(x0: number, y0: number, x1: number, y1: number, r: number, g: number, b: number, a: number) {
    this.ensure(6)
    this.push(x0, y0, r, g, b, a)
    this.push(x1, y0, r, g, b, a)
    this.push(x0, y1, r, g, b, a)
    this.push(x1, y0, r, g, b, a)
    this.push(x1, y1, r, g, b, a)
    this.push(x0, y1, r, g, b, a)
  }

  /** 垂直渐变矩形：顶部颜色 → 底部颜色（逐顶点插值，等价两端色标的线性渐变） */
  rectVGrad(
    x0: number, y0: number, x1: number, y1: number,
    r0: number, g0: number, b0: number, a0: number,
    r1: number, g1: number, b1: number, a1: number,
  ) {
    this.ensure(6)
    this.push(x0, y0, r0, g0, b0, a0)
    this.push(x1, y0, r0, g0, b0, a0)
    this.push(x0, y1, r1, g1, b1, a1)
    this.push(x1, y0, r0, g0, b0, a0)
    this.push(x1, y1, r1, g1, b1, a1)
    this.push(x0, y1, r1, g1, b1, a1)
  }

  /**
   * 线段 → 沿法线加宽的四边形（GL lineWidth 恒为 1，宽度必须几何化）。
   * 平头端：拖尾/轨迹由连续短段组成，段间缝隙可忽略；
   * round=true 时两端补半圆头（圆帽）——旗杆顶、虚线断点、折线轮廓的"首尾圆润"。
   */
  stroke(x0: number, y0: number, x1: number, y1: number, w: number, r: number, g: number, b: number, a: number, round = false) {
    const dx = x1 - x0
    const dy = y1 - y0
    const len = Math.sqrt(dx * dx + dy * dy)
    if (len < 1e-8) return
    const hw = w / 2 / len
    const nx = -dy * hw
    const ny = dx * hw
    this.ensure(6)
    this.push(x0 + nx, y0 + ny, r, g, b, a)
    this.push(x1 + nx, y1 + ny, r, g, b, a)
    this.push(x0 - nx, y0 - ny, r, g, b, a)
    this.push(x1 + nx, y1 + ny, r, g, b, a)
    this.push(x1 - nx, y1 - ny, r, g, b, a)
    this.push(x0 - nx, y0 - ny, r, g, b, a)
    if (round) {
      this.disc(x0, y0, w / 2, w / 2, 0, 8, r, g, b, a)
      this.disc(x1, y1, w / 2, w / 2, 0, 8, r, g, b, a)
    }
  }

  /** 斜接长度钳制系数（相对半宽）：过锐转角等效斜切，防尖刺 */
  private static MITER_LIMIT = 4

  /**
   * 折线描边：相邻线段在转角处按角平分线斜接（miter）相连，转角无缝。
   * 首尾平头、零长段跳过；要求各段法线同侧（地形等单调折线适用），折回轨迹仍用逐段 stroke。
   * pts 为 [x0,y0,x1,y1,...] 平铺，n 为浮点数个数（偶数）。
   */
  polyline(pts: Float32Array, n: number, w: number, r: number, g: number, b: number, a: number) {
    this.miter(pts, n, w, r, g, b, a)
  }

  /** 折线描边 + 逐顶点透明度（alpha[i] 对应 pts[2i] 处，段内线性渐变）：
   * 轨迹类线条按时间淡出时保持斜接无缝，避免按段平头四边形露角。 */
  polylineFade(pts: Float32Array, n: number, w: number, r: number, g: number, b: number, alpha: Float32Array) {
    this.miter(pts, n, w, r, g, b, alpha)
  }

  private miter(
    pts: Float32Array, n: number, w: number,
    r: number, g: number, b: number,
    alpha: Float32Array | number,
  ) {
    if (n < 4) return
    const hw = w / 2
    const limit = MeshBatch.MITER_LIMIT * hw
    const fade = typeof alpha !== 'number'
    // 上段状态：起点 (sx,sy)、起点端斜接向量 (mx,my)（已含半宽）、上段单位法线 (pnx,pny)
    let sx = 0
    let sy = 0
    let mx = 0
    let my = 0
    let pnx = 0
    let pny = 0
    let ready = false
    for (let i = 0; i + 3 < n; i += 2) {
      const ax = pts[i]
      const ay = pts[i + 1]
      const bx = pts[i + 2]
      const by = pts[i + 3]
      const dx = bx - ax
      const dy = by - ay
      const len = Math.sqrt(dx * dx + dy * dy)
      if (len < 1e-8) continue
      const nx = -dy / len
      const ny = dx / len
      if (!ready) {
        sx = ax
        sy = ay
        mx = nx * hw
        my = ny * hw
        pnx = nx
        pny = ny
        ready = true
        continue
      }
      // 本段起点 = 上段终点：斜接向量 = 两段法线之和（角平分线方向）
      let tx = pnx + nx
      let ty = pny + ny
      let fl = Math.sqrt(tx * tx + ty * ty)
      if (fl < 1e-6) {
        // 180° 折返：退化，按平头退避
        tx = nx
        ty = ny
        fl = 1
      } else {
        tx /= fl
        ty /= fl
        fl = hw / (pnx * tx + pny * ty) // = hw / cos(θ/2)
        if (fl > limit) fl = limit
      }
      const ex = tx * fl
      const ey = ty * fl
      const a0 = fade ? alpha[i / 2 - 1] : alpha
      const a1 = fade ? alpha[i / 2] : alpha
      this.ensure(6)
      this.push(sx + mx, sy + my, r, g, b, a0)
      this.push(ax + ex, ay + ey, r, g, b, a1)
      this.push(sx - mx, sy - my, r, g, b, a0)
      this.push(ax + ex, ay + ey, r, g, b, a1)
      this.push(ax - ex, ay - ey, r, g, b, a1)
      this.push(sx - mx, sy - my, r, g, b, a0)
      sx = ax
      sy = ay
      mx = ex
      my = ey
      pnx = nx
      pny = ny
    }
    if (ready) {
      const bx = pts[n - 2]
      const by = pts[n - 1]
      const ex = pnx * hw
      const ey = pny * hw
      const a0 = fade ? alpha[(n - 4) / 2] : alpha
      const a1 = fade ? alpha[(n - 2) / 2] : alpha
      this.ensure(6)
      this.push(sx + mx, sy + my, r, g, b, a0)
      this.push(bx + ex, by + ey, r, g, b, a1)
      this.push(sx - mx, sy - my, r, g, b, a0)
      this.push(bx + ex, by + ey, r, g, b, a1)
      this.push(bx - ex, by - ey, r, g, b, a1)
      this.push(sx - mx, sy - my, r, g, b, a0)
    }
  }

  /** 椭圆扇形填充：rot 为长轴旋转角（弧度）。rx = ry 时即正圆。 */
  disc(
    cx: number, cy: number, rx: number, ry: number, rot: number, seg: number,
    r: number, g: number, b: number, a: number,
  ) {
    if (rx <= 0 || ry <= 0 || a <= 0) return
    this.ensure(seg * 3)
    const cos = Math.cos(rot)
    const sin = Math.sin(rot)
    let px = 0
    let py = 0
    for (let i = 0; i <= seg; i++) {
      const th = (i / seg) * Math.PI * 2
      const ex = rx * Math.cos(th)
      const ey = ry * Math.sin(th)
      const qx = cx + ex * cos - ey * sin
      const qy = cy + ex * sin + ey * cos
      if (i > 0) this.tri(cx, cy, px, py, qx, qy, r, g, b, a)
      px = qx
      py = qy
    }
  }

  /**
   * 径向渐变圆盘：中心 → 边缘逐顶点线性插值，与 createRadialGradient 两端色标
   * 视觉等价，免去每帧建渐变对象或烘焙位图。
   */
  discGrad(
    cx: number, cy: number, radius: number, seg: number,
    cr: number, cg: number, cb: number, ca: number,
    er: number, eg: number, eb: number, ea: number,
  ) {
    if (radius <= 0) return
    this.ensure(seg * 3)
    let px = 0
    let py = 0
    for (let i = 0; i <= seg; i++) {
      const th = (i / seg) * Math.PI * 2
      const qx = cx + radius * Math.cos(th)
      const qy = cy + radius * Math.sin(th)
      if (i > 0) {
        this.ensure(3)
        this.push(cx, cy, cr, cg, cb, ca)
        this.push(px, py, er, eg, eb, ea)
        this.push(qx, qy, er, eg, eb, ea)
      }
      px = qx
      py = qy
    }
  }

  /**
   * 径向渐变圆盘：中心到 solidFrac·radius 保持实色，之外线性衰减到边缘色。
   * 雾状图元（云朵）的"实核 + 可控软边"：模糊半径 = (1-solidFrac)·radius，
   * 比 discGrad 的全径线性衰减边缘更利落。
   */
  discGradCore(
    cx: number, cy: number, radius: number, seg: number, solidFrac: number,
    cr: number, cg: number, cb: number, ca: number,
    er: number, eg: number, eb: number, ea: number,
  ) {
    if (radius <= 0) return
    this.ensure(seg * 9)
    const inner = radius * solidFrac
    let px = 0
    let py = 0
    let ix = 0
    let iy = 0
    for (let i = 0; i <= seg; i++) {
      const th = (i / seg) * Math.PI * 2
      const ex = cx + radius * Math.cos(th)
      const ey = cy + radius * Math.sin(th)
      const nx = cx + inner * Math.cos(th)
      const ny = cy + inner * Math.sin(th)
      if (i > 0) {
        this.push(cx, cy, cr, cg, cb, ca)
        this.push(ix, iy, cr, cg, cb, ca)
        this.push(nx, ny, cr, cg, cb, ca)
        this.push(ix, iy, cr, cg, cb, ca)
        this.push(px, py, er, eg, eb, ea)
        this.push(ex, ey, er, eg, eb, ea)
        this.push(ix, iy, cr, cg, cb, ca)
        this.push(ex, ey, er, eg, eb, ea)
        this.push(nx, ny, cr, cg, cb, ca)
      }
      px = ex
      py = ey
      ix = nx
      iy = ny
    }
  }

  ring(
    cx: number, cy: number, rx: number, ry: number, rot: number, seg: number, w: number,
    r: number, g: number, b: number, a: number,
  ) {
    if (rx <= 0 || ry <= 0 || a <= 0) return
    const cos = Math.cos(rot)
    const sin = Math.sin(rot)
    let px = 0
    let py = 0
    for (let i = 0; i <= seg; i++) {
      const th = (i / seg) * Math.PI * 2
      const ex = rx * Math.cos(th)
      const ey = ry * Math.sin(th)
      const qx = cx + ex * cos - ey * sin
      const qy = cy + ex * sin + ey * cos
      if (i > 0) this.stroke(px, py, qx, qy, w, r, g, b, a)
      px = qx
      py = qy
    }
  }

  /** 圆弧/虚线展开的采样点 scratch：按需扩容（最坏 = 弧线分段数 + 1） */
  private arcPts = new Float32Array(0)

  /** 圆弧描边：角度 a0 → a1（弧度，y 向下坐标系），线宽 w。
   * 相邻段斜接相连（转角无缝），弧线两端补圆头——虚线/圆弧首尾不再平头截断。 */
  arc(
    cx: number, cy: number, radius: number, a0: number, a1: number, seg: number, w: number,
    r: number, g: number, b: number, a: number,
  ) {
    if (radius <= 0 || a <= 0 || a1 === a0) return
    const n = seg + 1
    if (this.arcPts.length < n * 2) this.arcPts = new Float32Array(n * 2)
    const pts = this.arcPts
    for (let i = 0; i <= seg; i++) {
      const th = a0 + ((a1 - a0) * i) / seg
      pts[i * 2] = cx + radius * Math.cos(th)
      pts[i * 2 + 1] = cy + radius * Math.sin(th)
    }
    this.polyline(pts, n * 2, w, r, g, b, a)
    const hw = w / 2
    this.disc(pts[0], pts[1], hw, hw, 0, 8, r, g, b, a)
    this.disc(pts[seg * 2], pts[seg * 2 + 1], hw, hw, 0, 8, r, g, b, a)
  }

  /** 虚线圆：on/off 为弧长（世界单位），从角度 0 起按周长铺排 */
  dashRing(
    cx: number, cy: number, radius: number, on: number, off: number, w: number,
    r: number, g: number, b: number, a: number,
  ) {
    const circ = Math.PI * 2 * radius
    const period = on + off
    if (circ <= 0 || period <= 0) return
    let s = 0
    while (s < circ) {
      const segLen = Math.min(on, circ - s)
      const a0 = s / radius
      const a1 = (s + segLen) / radius
      const segs = Math.max(2, Math.ceil(segLen / 0.5))
      this.arc(cx, cy, radius, a0, a1, segs, w, r, g, b, a)
      s += period
    }
  }
}

File: sfgame-web/native/batch.c
// 渲染顶点批内核（C，替代原 AssemblyScript assembly/batch.ts）：x,y,r,g,b,a 平铺，f64 中间量 → f32 存储。
// 静态容量 + 静态内存：实例化时定型、运行期零分配、memory.buffer 视图恒定；
// 容量溢出整体丢弃图元（最坏场景 ~10 万顶点，容量留 ~2 倍余量）。
// 三角函数用 libm（musl 移植，与原 AS 的 musl 内核实现同源）；PI 用 hex float 保与旧版逐位一致。

#include <math.h>
#include <stddef.h>
#include <stdint.h>

#define VERTEX_STRIDE 6
#define CAPACITY 196608
#define PTS_CAP 2048
#define FADE_CAP 1024
#define MITER_LIMIT 4
#define PI 0x1.921FB54442D18p+1

static float data[CAPACITY * VERTEX_STRIDE];
static float ptsBuf[PTS_CAP];
static float fadeBuf[FADE_CAP];
static float gradRing[8];

static int count = 0;

int bCapacity(void) { return CAPACITY; }
int bPtsCap(void) { return PTS_CAP; }
int bFadeCap(void) { return FADE_CAP; }
uint32_t bData(void) { return (uint32_t)(uintptr_t)data; }
uint32_t bPtsBuf(void) { return (uint32_t)(uintptr_t)ptsBuf; }
uint32_t bFadeBuf(void) { return (uint32_t)(uintptr_t)fadeBuf; }
int bCount(void) { return count; }
void bReset(void) { count = 0; }

static void push(double x, double y, double r, double g, double b, double a) {
  size_t o = (size_t)count * VERTEX_STRIDE;
  data[o] = (float)x;
  data[o + 1] = (float)y;
  data[o + 2] = (float)r;
  data[o + 3] = (float)g;
  data[o + 4] = (float)b;
  data[o + 5] = (float)a;
  count++;
}

void bTri(
  double x0, double y0, double x1, double y1, double x2, double y2,
  double r, double g, double b, double a
) {
  if (count + 3 > CAPACITY) return;
  push(x0, y0, r, g, b, a);
  push(x1, y1, r, g, b, a);
  push(x2, y2, r, g, b, a);
}

void bRect(
  double x0, double y0, double x1, double y1,
  double r, double g, double b, double a
) {
  if (count + 6 > CAPACITY) return;
  push(x0, y0, r, g, b, a);
  push(x1, y0, r, g, b, a);
  push(x0, y1, r, g, b, a);
  push(x1, y0, r, g, b, a);
  push(x1, y1, r, g, b, a);
  push(x0, y1, r, g, b, a);
}

void bRectVGrad(
  double x0, double y0, double x1, double y1,
  double r0, double g0, double b0, double a0,
  double r1, double g1, double b1, double a1
) {
  if (count + 6 > CAPACITY) return;
  push(x0, y0, r0, g0, b0, a0);
  push(x1, y0, r0, g0, b0, a0);
  push(x0, y1, r1, g1, b1, a1);
  push(x1, y0, r0, g0, b0, a0);
  push(x1, y1, r1, g1, b1, a1);
  push(x0, y1, r1, g1, b1, a1);
}

static void disc(
  double cx, double cy, double rx, double ry, double rot, int seg,
  double r, double g, double b, double a
) {
  if (rx <= 0 || ry <= 0 || a <= 0) return;
  if (count + seg * 3 > CAPACITY) return;
  double c = cos(rot);
  double s = sin(rot);
  double px = 0, py = 0;
  for (int i = 0; i <= seg; i++) {
    double th = ((double)i / (double)seg) * PI * 2;
    double ex = rx * cos(th);
    double ey = ry * sin(th);
    double qx = cx + ex * c - ey * s;
    double qy = cy + ex * s + ey * c;
    if (i > 0) {
      push(cx, cy, r, g, b, a);
      push(px, py, r, g, b, a);
      push(qx, qy, r, g, b, a);
    }
    px = qx;
    py = qy;
  }
}

void bDisc(
  double cx, double cy, double rx, double ry, double rot, int seg,
  double r, double g, double b, double a
) {
  disc(cx, cy, rx, ry, rot, seg, r, g, b, a);
}

static void stroke(
  double x0, double y0, double x1, double y1, double w,
  double r, double g, double b, double a, int round
) {
  double dx = x1 - x0;
  double dy = y1 - y0;
  double len = sqrt(dx * dx + dy * dy);
  if (len < 1e-8) return;
  if (count + (round ? 54 : 6) > CAPACITY) return;
  double hw = w / 2 / len;
  double nx = -dy * hw;
  double ny = dx * hw;
  push(x0 + nx, y0 + ny, r, g, b, a);
  push(x1 + nx, y1 + ny, r, g, b, a);
  push(x0 - nx, y0 - ny, r, g, b, a);
  push(x1 + nx, y1 + ny, r, g, b, a);
  push(x1 - nx, y1 - ny, r, g, b, a);
  push(x0 - nx, y0 - ny, r, g, b, a);
  if (round) {
    disc(x0, y0, w / 2, w / 2, 0, 8, r, g, b, a);
    disc(x1, y1, w / 2, w / 2, 0, 8, r, g, b, a);
  }
}

void bStroke(
  double x0, double y0, double x1, double y1, double w,
  double r, double g, double b, double a, int round
) {
  stroke(x0, y0, x1, y1, w, r, g, b, a, round);
}

// pts/fade 由宿主写入 ptsBuf/fadeBuf 后调用；n 为 float 个数（点数 × 2）
static void miter(int n, double w, double r, double g, double b, double baseA, int fade) {
  if (n < 4) return;
  if (count + 3 * n > CAPACITY) return;
  double hw = w / 2;
  double limit = MITER_LIMIT * hw;
  double sx = 0, sy = 0, mx = 0, my = 0, pnx = 0, pny = 0;
  int ready = 0;
  for (int i = 0; i + 3 < n; i += 2) {
    double ax = (double)ptsBuf[i];
    double ay = (double)ptsBuf[i + 1];
    double bx = (double)ptsBuf[i + 2];
    double by = (double)ptsBuf[i + 3];
    double dx = bx - ax;
    double dy = by - ay;
    double len = sqrt(dx * dx + dy * dy);
    if (len < 1e-8) continue;
    double nx = -dy / len;
    double ny = dx / len;
    if (!ready) {
      sx = ax;
      sy = ay;
      mx = nx * hw;
      my = ny * hw;
      pnx = nx;
      pny = ny;
      ready = 1;
      continue;
    }
    double tx = pnx + nx;
    double ty = pny + ny;
    double fl = sqrt(tx * tx + ty * ty);
    if (fl < 1e-6) {
      tx = nx;
      ty = ny;
      fl = 1;
    } else {
      tx /= fl;
      ty /= fl;
      fl = hw / (pnx * tx + pny * ty);
      if (fl > limit) fl = limit;
    }
    double ex = tx * fl;
    double ey = ty * fl;
    double a0 = fade ? (double)fadeBuf[i / 2 - 1] : baseA;
    double a1 = fade ? (double)fadeBuf[i / 2] : baseA;
    push(sx + mx, sy + my, r, g, b, a0);
    push(ax + ex, ay + ey, r, g, b, a1);
    push(sx - mx, sy - my, r, g, b, a0);
    push(ax + ex, ay + ey, r, g, b, a1);
    push(ax - ex, ay - ey, r, g, b, a1);
    push(sx - mx, sy - my, r, g, b, a0);
    sx = ax;
    sy = ay;
    mx = ex;
    my = ey;
    pnx = nx;
    pny = ny;
  }
  if (ready) {
    double bx = (double)ptsBuf[n - 2];
    double by = (double)ptsBuf[n - 1];
    double ex = pnx * hw;
    double ey = pny * hw;
    double a0 = fade ? (double)fadeBuf[(n - 4) / 2] : baseA;
    double a1 = fade ? (double)fadeBuf[(n - 2) / 2] : baseA;
    push(sx + mx, sy + my, r, g, b, a0);
    push(bx + ex, by + ey, r, g, b, a1);
    push(sx - mx, sy - my, r, g, b, a0);
    push(bx + ex, by + ey, r, g, b, a1);
    push(bx - ex, by - ey, r, g, b, a1);
    push(sx - mx, sy - my, r, g, b, a0);
  }
}

void bPolyline(int n, double w, double r, double g, double b, double a) {
  miter(n, w, r, g, b, a, 0);
}

void bPolylineFade(int n, double w, double r, double g, double b) {
  miter(n, w, r, g, b, 0, 1);
}

void bDiscGrad(
  double cx, double cy, double radius, int seg,
  double cr, double cg, double cb, double ca,
  double er, double eg, double eb, double ea
) {
  if (radius <= 0) return;
  if (count + seg * 3 > CAPACITY) return;
  double px = 0, py = 0;
  for (int i = 0; i <= seg; i++) {
    double th = ((double)i / (double)seg) * PI * 2;
    double qx = cx + radius * cos(th);
    double qy = cy + radius * sin(th);
    if (i > 0) {
      push(cx, cy, cr, cg, cb, ca);
      push(px, py, er, eg, eb, ea);
      push(qx, qy, er, eg, eb, ea);
    }
    px = qx;
    py = qy;
  }
}

void bDiscGradCore(
  double cx, double cy, double radius, int seg, double solidFrac,
  double cr, double cg, double cb, double ca,
  double er, double eg, double eb, double ea
) {
  if (radius <= 0) return;
  if (count + seg * 21 > CAPACITY) return;
  double inner = radius * solidFrac;
  double band = radius - inner;
  double r1 = inner + band / 3;
  double r2 = inner + (band * 2) / 3;
  double t1 = 7.0 / 27;
  double t2 = 20.0 / 27;
  double c1r = cr + (er - cr) * t1, c1g = cg + (eg - cg) * t1, c1b = cb + (eb - cb) * t1, c1a = ca + (ea - ca) * t1;
  double c2r = cr + (er - cr) * t2, c2g = cg + (eg - cg) * t2, c2b = cb + (eb - cb) * t2, c2a = ca + (ea - ca) * t2;
  for (int i = 0; i <= seg; i++) {
    double th = ((double)i / (double)seg) * PI * 2;
    double cs = cos(th);
    double sn = sin(th);
    double n0x = cx + inner * cs, n0y = cy + inner * sn;
    double n1x = cx + r1 * cs, n1y = cy + r1 * sn;
    double n2x = cx + r2 * cs, n2y = cy + r2 * sn;
    double n3x = cx + radius * cs, n3y = cy + radius * sn;
    if (i > 0) {
      push(cx, cy, cr, cg, cb, ca);
      push((double)gradRing[0], (double)gradRing[1], cr, cg, cb, ca);
      push(n0x, n0y, cr, cg, cb, ca);
      push((double)gradRing[0], (double)gradRing[1], cr, cg, cb, ca);
      push((double)gradRing[2], (double)gradRing[3], c1r, c1g, c1b, c1a);
      push(n1x, n1y, c1r, c1g, c1b, c1a);
      push((double)gradRing[0], (double)gradRing[1], cr, cg, cb, ca);
      push(n1x, n1y, c1r, c1g, c1b, c1a);
      push(n0x, n0y, cr, cg, cb, ca);
      push((double)gradRing[2], (double)gradRing[3], c1r, c1g, c1b, c1a);
      push((double)gradRing[4], (double)gradRing[5], c2r, c2g, c2b, c2a);
      push(n2x, n2y, c2r, c2g, c2b, c2a);
      push((double)gradRing[2], (double)gradRing[3], c1r, c1g, c1b, c1a);
      push(n2x, n2y, c2r, c2g, c2b, c2a);
      push(n1x, n1y, c1r, c1g, c1b, c1a);
      push((double)gradRing[4], (double)gradRing[5], c2r, c2g, c2b, c2a);
      push((double)gradRing[6], (double)gradRing[7], er, eg, eb, ea);
      push(n3x, n3y, er, eg, eb, ea);
      push((double)gradRing[4], (double)gradRing[5], c2r, c2g, c2b, c2a);
      push(n3x, n3y, er, eg, eb, ea);
      push(n2x, n2y, c2r, c2g, c2b, c2a);
    }
    gradRing[0] = (float)n0x;
    gradRing[1] = (float)n0y;
    gradRing[2] = (float)n1x;
    gradRing[3] = (float)n1y;
    gradRing[4] = (float)n2x;
    gradRing[5] = (float)n2y;
    gradRing[6] = (float)n3x;
    gradRing[7] = (float)n3y;
  }
}

void bRing(
  double cx, double cy, double rx, double ry, double rot, int seg, double w,
  double r, double g, double b, double a
) {
  if (rx <= 0 || ry <= 0 || a <= 0) return;
  if (count + seg * 6 > CAPACITY) return;
  double c = cos(rot);
  double s = sin(rot);
  double px = 0, py = 0;
  for (int i = 0; i <= seg; i++) {
    double th = ((double)i / (double)seg) * PI * 2;
    double ex = rx * cos(th);
    double ey = ry * sin(th);
    double qx = cx + ex * c - ey * s;
    double qy = cy + ex * s + ey * c;
    if (i > 0) stroke(px, py, qx, qy, w, r, g, b, a, 0);
    px = qx;
    py = qy;
  }
}

// 与 polyline 复用 ptsBuf：调用方不得在 arc 期间改写入参缓冲（单线程天然成立）
static void arc(
  double cx, double cy, double radius, double a0, double a1, int seg, double w,
  double r, double g, double b, double a
) {
  if (radius <= 0 || a <= 0 || a1 == a0) return;
  int n = seg + 1;
  if (n * 2 > PTS_CAP) return;
  for (int i = 0; i <= seg; i++) {
    double th = a0 + ((a1 - a0) * (double)i) / (double)seg;
    ptsBuf[i * 2] = (float)(cx + radius * cos(th));
    ptsBuf[i * 2 + 1] = (float)(cy + radius * sin(th));
  }
  miter(n * 2, w, r, g, b, a, 0);
  double hw = w / 2;
  disc((double)ptsBuf[0], (double)ptsBuf[1], hw, hw, 0, 8, r, g, b, a);
  disc((double)ptsBuf[seg * 2], (double)ptsBuf[seg * 2 + 1], hw, hw, 0, 8, r, g, b, a);
}

void bArc(
  double cx, double cy, double radius, double a0, double a1, int seg, double w,
  double r, double g, double b, double a
) {
  arc(cx, cy, radius, a0, a1, seg, w, r, g, b, a);
}

void bDashRing(
  double cx, double cy, double radius, double on, double off, double w,
  double r, double g, double b, double a
) {
  double circ = PI * 2 * radius;
  double onLen = on;
  double offLen = off;
  double period = onLen + offLen;
  if (circ <= 0 || period <= 0) return;
  if (circ < 6 * period) {
    double k = circ / (6 * period);
    onLen *= k;
    offLen *= k;
    period = onLen + offLen;
  }
  double s = 0;
  while (s < circ) {
    double segLen = fmin(onLen, circ - s);
    double ang0 = s / radius;
    double ang1 = (s + segLen) / radius;
    int segs = (int)ceil(segLen / 0.5);
    if (segs < 2) segs = 2;
    arc(cx, cy, radius, ang0, ang1, segs, w, r, g, b, a);
    s += period;
  }
}
